### PR TITLE
Send and Receive

### LINKED ISSUE #4: Initial FFi wrapper for Send
Implement Send methods and expose in ffi wrapper. This may not be final and may require error handling and some additional tweaks however this should implement all Send methods from rust-payjoin

### LINKED ISSUE #5: Initial FFI wrapper for Receive
Implement Recieve methods and expose in ffi wrapper. This may not be final and may require error handling and some additional tweaks however this should implement all Receive methods from rust-payjoin

### PR DESCRIPTION
**Pull request by BitcoinZavior**
- [x] Initial build script
- [x] Initial Error handling
- [x] Send
- [x] Receive
- [x] Replicate rust-payjoin test cases

Fixes #4 and Fixes #5

### CODE REVIEW

**Comment by BitcoinZavior** [comment]
Please review @StaxoLotl @DanGould

**Comment by StaxoLotl** [comment]
Coming along well, will take a look again when the build scripts are added

File: src/receive.rs [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+
+	/// Call after checking that the Original PSBT can be broadcast.
+	///
+	/// Receiver MUST check that the Original PSBT from the sender can be broadcast, i.e. testmempoolaccept bitcoind rpc returns { “allowed”: true,.. } for get_transaction_to_check_broadcast() before calling this method.
+	///
+	/// Do this check if you generate bitcoin uri to receive Payjoin on sender request without manual human approval, like a payment processor. Such so called “non-interactive” receivers are otherwise vulnerable to probing attacks. If a sender can make requests at will, they can learn which bitcoin the receiver owns at no cost. Broadcasting the Original PSBT after some time in the failure case makes incurs sender cost and prevents probing.
+	///
+	/// Call this after checking downstream.
+	pub fn check_can_broadcast(
+		self, can_broadcast: Box<dyn CanBroadcast>,
```
Comment (DanGould) [inline src/receive.rs:69]: why is `CanBroadcast` a trait rather than a function that returns a bool like the others? Also, I understand that bitcoin core rpc expects a Vec<String>, but I think this function really only requires a single tx_hex: String as a generic.

File: src/receive.rs [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+	) -> Result<Self, anyhow::Error> {
+		match PdkUncheckedProposal::from_request(body, query.as_str(), headers) {
+			Ok(e) => Ok(UncheckedProposal { internal: e }),
+			Err(e) => anyhow::bail!("{e}"),
+		}
+	}
+
+	/// The Sender’s Original PSBT
+	pub fn get_transaction_to_schedule_broadcast(&self) -> Transaction {
+		let res = self.internal.get_transaction_to_schedule_broadcast();
```
Comment (DanGould) [inline src/receive.rs:57]: get_ prefix is not typically used in rust code. I guess this needs to be changed upstream!

https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter

File: src/receive.rs [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+		}) {
+			Ok(e) => Ok(PayjoinProposal { internal: Mutex::new(Some(e)) }),
+			Err(e) => anyhow::bail!("{e}"),
+		}
+	}
+}
+
+/// A mutable checked proposal that the receiver may contribute inputs to to make a payjoin.
+pub struct PayjoinProposal {
+	pub internal: Mutex<Option<PdkPayjoinProposal>>,
```
Comment (DanGould) [inline src/receive.rs:195]: Why is a Mutex introduced here? Seems like that adds additional complexity? There should be no situation where this struct exists but the proposal is not initiated since it has no constructor and can only be created from calling the prior type's check.

File: src/receive.rs [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+}
+impl PayjoinProposal {
+	pub(crate) fn get_proposal(&self) -> MutexGuard<Option<PdkPayjoinProposal>> {
+		self.internal.lock().expect("PayjoinProposal")
+	}
```
Comment (DanGould) [inline src/receive.rs:200]: Seems like this method could not exist if there were not Mutex

File: src/send.rs [hunk 5]
```diff
@@ -1,74 +1,130 @@
...
+use std::sync::{Arc, Mutex, MutexGuard};
 
 ///Builder for sender-side payjoin parameters
+///
 ///These parameters define how client wants to handle Payjoin.
 
-pub(crate) struct Configuration {
-    pub(crate) internal: PdkConfiguration,
+pub struct Configuration {
+	pub internal: Mutex<Option<PdkConfiguration>>,
```
Comment (DanGould) [inline src/send.rs:11]: Again, why is a Mutex necessary? When would this struct exist where internal = None?

File: src/receive.rs [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+			{
+				let network = Network::Bitcoin;
+				Ok(Address::from_script(script, network.clone()).unwrap()
+					== Address::from_str("3CZZi7aWFugaCdUCS15dgrUUViupmB8bVM")
+						.unwrap()
+						.assume_checked()
+						.unwrap())
+			}
+		}
+	}
```
Comment (DanGould) [inline src/receive.rs:344]: This smells like an antipattern. We should be able to pass functions pointers instead of implementing traits, no?

**Review by DanGould — COMMENTED** [review]
Absolutely incredible work that passes tests on my machine. This review helped me find a few issues upstream to fix too. The rust-payjoin master branch has a slightly different receive flow around fees and payload preparation I'll have to package into a release which should reduce some of the binding surface area too.

My main concern is around passing function pointers, which I thought in a conversation with @thunderbiscuit we concluded was possible. It would be much better to do that, as in the existing implementation tests, rather than require downstream trait implementations.

Passing function pointers would allow us to keep the typestate engine and get rid of the uninitialized / Mutex struct problem.

Truly excellent to see this effort moving forward!

**Comment by BitcoinZavior** [comment]
> My main concern is around passing function pointers, which I thought in a conversation with @thunderbiscuit we concluded was possible. It would be much better to do that, as in the existing implementation tests, rather than require downstream trait implementations.
> 
> Passing function pointers would allow us to keep the typestate engine and get rid of the uninitialized / Mutex struct problem.
> 
Passing callbacks is defenitly possible in rust, but is not recomened by uniffi, the recomended Uniffi approach is to use Traits. Ref: https://mozilla.github.io/uniffi-rs/udl/callback_interfaces.html? highlight=callback#using-callback-interfaces

File: src/receive.rs (hunk 1 above)
Comment (BitcoinZavior) [reply src/receive.rs:69]: Because using CanBroadcast trait is the recommended way to use callbacks when we use uniffi for generating code (But can look into using function callback instead)

> Also, I understand that bitcoin core rpc expects a Vec, but I think this function really only 
> requires a single tx_hex: String as a generic.
I got inspiration from the test code [integration.rs](http://integration.rs/)
```
let proposal = proposal
       .check_can_broadcast(|tx| {
         Ok(receiver
           .test_mempool_accept(&[bitcoin::consensus::encode::serialize_hex(&tx)])
           .unwrap()
           .first()
           .unwrap()
           .allowed)
       })
       .expect("Payjoin proposal should be broadcastable");`
```
But I think I can change this to a Vec.

File: src/receive.rs (hunk 3 above)
Comment (BitcoinZavior) [reply src/receive.rs:195]: There are a few reasons to use a mutexGaurd
**PayjoinProposal.**
 This is the `PayjoinProposal` struct without mutex_gaurd,
```
pub struct PayjoinProposal {
   internal: PdkPayjoinProposal,
}
impl PayjoinProposal {
   pub fn contribute_witness_input(&mut self, txout: TxOut, outpoint: OutPoint) {
    // let mut guard = self.get_proposal_mutex_guard();
    self.internal.contribute_witness_input(txout.into(), outpoint.into());
    // [guard.as](http://guard.as/)_mut().unwrap().contribute_witness_input(txout.into(), outpoint.into());
   }
}
```
    1, In the above function we have to use `&mut self` in the parameter,
      but unfortunetly, `uniffi` , doesn't allow that.
    2, prepare psbt takes a `mut self`, that will be a problem as well,
      since uniffi requires a `&self` in the function for properly generating
      binding code.
In short  mutuxes are used to ensure safe and synchronised access to shared resources in multi-threaded or multi-process programs. 
But good to keep checking, I am sure there are instance where we can do without a mutex.
FYI I started off with tons more mutex' and removed after some analysis. If you come across any which you are sure dont need ro be a mutex please point it out.

File: src/receive.rs (hunk 4 above)
Comment (BitcoinZavior) [reply src/receive.rs:200]: Yes you are right, but this is more of a helper function.

File: src/send.rs (hunk 5 above)
Comment (BitcoinZavior) [reply src/send.rs:11]: Same as before, yes we can remove it if we feel mutex is unnecessary here.

File: src/receive.rs (hunk 6 above)
Comment (BitcoinZavior) [reply src/receive.rs:344]: We won't able to pass functions pointers directly in payjoin_ffi.udl file because of uniffi. Using traits is the option we have at the moment. Might be more options to deal with this in the future.

File: src/lib.rs [hunk 7]
```diff
@@ -1,121 +1,206 @@
...
+}
+
+//TODO; RECREATE ADDRESS STRUCTURE
+#[derive(Debug, Clone, PartialEq)]
+pub struct Address<T>
+where
+	T: bitcoin::address::NetworkValidation,
+{
+	pub internal: _BitcoinAdrress<T>,
+}
```
[the remark this answers is not in the document]
Comment (BitcoinZavior) [reply src/lib.rs:48]: This should be removed, I will remove it in a future commit.

**Comment by BitcoinZavior** [comment]
I am merging this branch, I have opened issues for the un addressed comments.
If there are any other issues we can open issues and work through them.

File: src/receive.rs (hunk 3 above)
Comment (DanGould) [reply src/receive.rs:195]: Could the mutex be removed and this simplified by passing an owned `self` instead of a mutable reference?

Comment (BitcoinZavior) [reply src/receive.rs:195]: Yes, if we dont want to use mutex here we can remove it. probably worth looking if there are other instances where mutex is not required. I will updated in a future commit in the python branch.
